src/components/Modals/Tasks/EditTask.jsx

```diff
@@ -1,159 +1,114 @@
 import PropTypes from 'prop-types';
-import React from 'react';
+import { useEffect, useReducer, useRef, useState } from 'react';
 import classes from './EditTask.module.css';
 import Button from '../../Button/Button';
 import CraftInput from '../CraftInput';
+import { stateReducer, TASK_ONCHANGE, TASK_VALIDATE, validatorReducer } from '../modals-helpers';
+import { checkAllFormValidity } from '../../../utilities/form-validators';
 
-import { EDIT_TASK_ONCHANGE, EDIT_TASK_VALIDATE_FIELDS, EDIT_TASK_VALIDATE_FORM, reducerFunc } from './task-helpers';
-import debounce from '../../../utilities/debounce';
+export default function EditTask({ closeFunc, task, liftUpEditTask, usersList }) {
+  const onChangeSnapshot = useRef({});
+  const firstRender = useRef(true);
 
-export default class EditTask extends React.PureComponent {
-  constructor(props) {
-    super(props);
-    this.state = {
-      data: {
-        title: '',
-        description: '',
-        startDate: '',
-        deadLine: '',
-        selectedUsers: [],
-      },
-      validator: {
-        title: true,
-        startDate: true,
-        deadLine: true,
-        selectedUsers: true,
-      },
-      errors: {
-        titleError: '',
-        startDateError: '',
-        deadLineError: '',
-        selectedUsersError: '',
-      },
-      usersList: [],
-      isValid: false,
-    };
-    this.onChange = this.onChange.bind(this);
-    this.liftUpEditTask = this.liftUpEditTask.bind(this);
-    this.closeModal = this.closeModal.bind(this);
-    this.validateFields = this.validateFields.bind(this);
-    this.validateForm = this.validateForm.bind(this);
-  }
+  const [state, dispatchState] = useReducer(stateReducer, undefined, () => ({
+    ...task,
+  }));
 
-  componentDidMount() {
-    const { task, usersList } = this.props;
-    this.setState((prevState) => ({
-      ...prevState,
-      data: {
-        ...task,
-      },
-      usersList,
-    }));
-  }
+  const [validator, dispatchValidator] = useReducer(validatorReducer, undefined, () => ({
+    validator: {
+      title: true,
+      startDate: true,
+      deadLine: true,
+      selectedUsers: true,
+    },
+    errors: {
+      titleError: '',
+      startDateError: '',
+      deadLineError: '',
+      selectedUsersError: '',
+    },
+  }));
 
-  onChange(event) {
-    const { target } = event;
-    const { name, value } = target;
-    const targetType = target.type;
-    this.setState(
-      (prevState) =>
-        reducerFunc(prevState, {
-          type: EDIT_TASK_ONCHANGE,
-          name,
-          value,
-          targetType,
-        }),
-      debounce(() => {
-        this.validateFields(name, value, targetType);
-      }, 1000),
-    );
-  }
+  useEffect(() => {
+    if (firstRender.current) {
+      firstRender.current = false;
```

Comment:
why do you need it ? it looks like your implementation is wrong if you need this in your case

Reply:
Reworked with creating custom hook:
```suggestion
export const useAllSelectedFormsValidityChecker = (validator, initialState) => {
  const [isValid, setIsValid] = useState(false);
  const initialStateRef = useRef(initialState);

  const isOneOfStateFieldsChanged = useMemo(() => checkIfOneOfFieldsChanged(initialStateRef.current, initialState), [
    initialStateRef.current,
    initialState,
  ]);
  useEffect(() => {
    if (isOneOfStateFieldsChanged) {
      console.log('useEffect setIsValid');
      setIsValid(checkAllFormValidity(validator));
    }
  }, [validator, isOneOfStateFieldsChanged]);

  return isValid;
};
```